test/libsolidity/semanticTests/events/event_anonymous_with_topics.sol

```diff
@@ -0,0 +1,11 @@
+contract ClientReceipt {
+    event Deposit(address indexed _from, bytes32 indexed _id, uint indexed _value, uint indexed _value2, bytes32 data) anonymous;
+    function deposit(bytes32 _id) public payable {
+        emit Deposit(msg.sender, _id, msg.value, 2, "abc");
+    }
+}
+// ====
+// compileViaYul: also
+// ----
+// deposit(bytes32), 18 wei: 0x1234 ->
+// ~ emit <anonymous>: #0x1212121212121212121212121212120000000012, #0x1234, #0x12, #0x02, 0x6162630000000000000000000000000000000000000000000000000000000000
```

Comment:
Can you add a similar test but with an anonymous event being emitted with the value of the first indexed parameter matching the hash of the signature of an existing event?

Reply:
If I understood you correctly, you want to check how the behaviour is in case of a potential signature hash collision of `topic[0]`. The thing is that for anonymous events `topic[0]` is not generated (see e.g. https://docs.soliditylang.org/en/v0.8.4/abi-spec.html?highlight=events#events "If the event is declared as anonymous the topics[0] is not generated") I added a test for this, but maybe it is not that meaningful. Did I understood you correctly?

Reply:
I'm not sure whether it makes sense to add this test. 
```solidity
contract ClientReceipt {
    event Deposit(uint256 indexed _from, bytes32 indexed _id, uint _value) anonymous;
    function deposit(bytes32 _id) public payable {
        emit Deposit(0x2012159c, _id, msg.value); // 0x2012159c -> 'Deposit(uint256,bytes32,uint256)'
    }
}
// ====
// compileViaYul: also
// ----
// deposit(bytes32), 18 wei: 0x1234 ->
// ~ emit <anonymous>: #0x2012159c, #0x1234, 0x12
```
I will not add this test for now, but let me know whether you still think it makes sense. Maybe I did not understood you correctly.

Reply:
It's related to this: #11439.

> If I understood you correctly, you want to check how the behaviour is in case of a potential signature hash collision of topic[0].

Yeah. I think you can misdetect an anonymous event as a different, non-anonymous one. And it's not really a collision like in a hash function where you have to find the preimage. You can just insert the hash directly into that parameter (as long as it's a value type) so spoofing another event is trivial.

I requested this test while still was not sure whether you can spoof events like that so I mostly wanted to see what will happen. Now I know you can but I'd still add one just to have this corner case covered.

I think that the snippet you gave above does not reproduce it because you used only the first 4 bytes of the hash. For events the whole hash is used.

> The thing is that for anonymous events topic[0] is not generated

I think it just means that the hash of the signature is not automatically generated and therefore not inserted into `topic[0]`. But `topic[0]` still contains something - one of the topics (probably the first one but I'm not sure if the extra fourth one is not just placed there instead). If that topic turns out to have the same value as the hash of the event signature, the anonymous event will be mistaken for a non-anonymous one. Also in that case the parameter types will likely not match.

I'll update this part in #11439 to make this clearer.

Reply:
Ok, looks like that I understood it differently. Let me check..

> It's related to this: #11439.
> 
> I think that the snippet you gave above does not reproduce it because you used only the first 4 bytes of the hash. For events the whole hash is used.

Thats true. :)

Reply:
Ok, I added a test `event_anonymous_with_signature_collision.sol`. I added a check within ~`SemanticTest::findEvent`~ `SemanticTest::matchEvent`, that only non-anonymous events can be matched.